File: leetcode.js/找出字符串中所有字母异位词.js
var findAnagrams = function(s, p) {
    const len_s = s.length;
    const len_p = p.length;
    if (len_s < len_p) {
        return [];
    }
    const result = [];
    const count_s = new Array(26).fill(0);
    const count_p = new Array(26).fill(0);
    
    // 初始化p的字符计数
    for (let i = 0; i < len_p; i++) {
        count_p[p.charCodeAt(i) - 'a'.charCodeAt(0)]++;
    }
    
    // 初始化s的第一个窗口计数
    for (let i = 0; i < len_p; i++) {
        count_s[s.charCodeAt(i) - 'a'.charCodeAt(0)]++;
    }
    
    // 优化数组比较：直接在函数内完成，避免函数调用开销
    const isEqual = () => {
        for (let i = 0; i < 26; i++) {
            if (count_s[i] !== count_p[i]) {
                return false;
            }
        }
        return true;
    };
    
    // 检查初始窗口
    if (isEqual()) {
        result.push(0);
    }
    
    // 滑动窗口移动
    for (let i = len_p; i < len_s; i++) {
        // 移除左边界字符（窗口外的字符）
        const leftChar = s.charCodeAt(i - len_p) - 'a'.charCodeAt(0);
        count_s[leftChar]--;
        
        // 添加右边界字符（新进入窗口的字符）
        const rightChar = s.charCodeAt(i) - 'a'.charCodeAt(0);
        count_s[rightChar]++;
        
        // 检查当前窗口
        if (isEqual()) {
            result.push(i - len_p + 1);
        }
    }
    
    return result;
};
// 实例调用
const s = "cbaebabacd";
const p = "abc";
const anagramIndices = findAnagrams(s, p);
console.log(anagramIndices); // 输出: [0, 6]

File: leetcode.py/找出字符串中所有字母异位词.py
class Solution(object):
    def findAnagrams(self, s, p):
        len_s = len(s)
        len_p = len(p)
        if len_s < len_p:
            return []

        result = []
        # 初始化计数数组
        count_s = [0] * 26
        count_p = [0] * 26

        # 统计 p 中字符的出现次数
        for char in p:
            count_p[ord(char) - ord('a')] += 1

        # 统计 s 中第一个滑动窗口内字符的出现次数
        for i in range(len_p):
            count_s[ord(s[i]) - ord('a')] += 1

        if count_s == count_p:
            result.append(0)

        # 滑动窗口移动
        for i in range(len_p, len_s):
            # 移除左边界的字符
            count_s[ord(s[i - len_p]) - ord('a')] -= 1
            # 添加右边界的字符
            count_s[ord(s[i]) - ord('a')] += 1
            if count_s == count_p:
                result.append(i - len_p + 1)

        return result
# 以下是使用 Solution 类的 findAnagrams 方法的示例
if __name__ == "__main__":
    solution = Solution()
    s = "cbaebabacd"
    p = "abc"
    print(solution.findAnagrams(s, p))  # 输出: [0, 6]
